File: openmrs-client/src/main/java/org/openmrs/mobile/activities/settings/SettingsFragment.java
/*
 * The contents of this file are subject to the OpenMRS Public License
 * Version 1.0 (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 * http://license.openmrs.org
 *
 * Software distributed under the License is distributed on an "AS IS"
 * basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See the
 * License for the specific language governing rights and limitations
 * under the License.
 *
 * Copyright (C) OpenMRS, LLC.  All Rights Reserved.
 */


package org.openmrs.mobile.activities.settings;

import android.content.ActivityNotFoundException;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.LinearLayout;
import android.widget.TextView;

import org.openmrs.mobile.R;
import org.openmrs.mobile.activities.ACBaseFragment;
import org.openmrs.mobile.activities.logs.LogsActivity;
import org.openmrs.mobile.services.ConceptDownloadService;
import org.openmrs.mobile.utilities.ApplicationConstants;
import org.openmrs.mobile.utilities.ToastUtil;

import java.util.Objects;

import androidx.annotation.NonNull;
import androidx.appcompat.widget.SwitchCompat;
import androidx.localbroadcastmanager.content.LocalBroadcastManager;

public class SettingsFragment extends ACBaseFragment<SettingsContract.Presenter> implements SettingsContract.View {

    private BroadcastReceiver bReceiver;

    private TextView conceptsInDbTextView;
    private ImageButton downloadConceptsButton;
    private SwitchCompat darkModeSwitch;

    private View root;

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        root = inflater.inflate(R.layout.fragment_settings, container, false);

        bReceiver = new BroadcastReceiver() {
            @Override
            public void onReceive(Context context, Intent intent) {
                mPresenter.updateConceptsInDBTextView();
            }
        };

        setUpConceptsView();
        return root;
    }

    @Override
    public void onPause() {
        super.onPause();
        LocalBroadcastManager.getInstance(this.getActivity()).unregisterReceiver(bReceiver);
    }

    @Override
    public void onResume() {
        super.onResume();
        mPresenter.updateConceptsInDBTextView();
        LocalBroadcastManager.getInstance(this.getActivity()).registerReceiver(bReceiver, new IntentFilter(ApplicationConstants.BroadcastActions.CONCEPT_DOWNLOAD_BROADCAST_INTENT_ID));
    }

    @Override
    public void setConceptsInDbText(String text) {
        if(text.equals("0")){
            downloadConceptsButton.setEnabled(true);
            ToastUtil.showLongToast(getActivity(),
                    ToastUtil.ToastType.WARNING,
                    R.string.settings_no_concepts_toast);
        }else{
            downloadConceptsButton.setEnabled(false);
        }
        conceptsInDbTextView.setText(text);
    }

    @Override
    public void addLogsInfo(long logSize, String logFilename) {
        LinearLayout logsLl = root.findViewById(R.id.frag_settings_logs_ll);
        TextView logsDesc1Tv = root.findViewById(R.id.frag_settings_logs_desc1_tv);
        TextView logsDesc2Tv = root.findViewById(R.id.frag_settings_logs_desc2_tv);

        logsDesc1Tv.setText(logFilename);
        logsDesc2Tv.setText(getContext().getString(R.string.settings_frag_size) + logSize + "kB");
        logsLl.setOnClickListener(view ->{
            Intent i = new Intent(view.getContext() , LogsActivity.class);
            startActivity(i);
        });
    }

    @Override
    public void setUpConceptsView() {
        conceptsInDbTextView = root.findViewById(R.id.frag_settings_concepts_count_tv);

        downloadConceptsButton = root.findViewById(R.id.frag_settings_concepts_download_btn);

        downloadConceptsButton.setOnClickListener(view -> {
            downloadConceptsButton.setEnabled(false);
            Intent startIntent = new Intent(getActivity(), ConceptDownloadService.class);
            startIntent.setAction(ApplicationConstants.ServiceActions.START_CONCEPT_DOWNLOAD_ACTION);
            Objects.requireNonNull(getActivity()).startService(startIntent);
        });
    }

    @Override
    public void addBuildVersionInfo() {
        String versionName = "";
        int buildVersion = 0;

        PackageManager packageManager = this.getActivity().getPackageManager();
        String packageName = this.getActivity().getPackageName();

        try {
            versionName = packageManager.getPackageInfo(packageName, 0).versionName;
            ApplicationInfo ai = packageManager.getApplicationInfo(packageName, PackageManager.GET_META_DATA);
            buildVersion = ai.metaData.getInt("buildVersion");
        } catch (PackageManager.NameNotFoundException e) {
            mPresenter.logException("Failed to load meta-data, NameNotFound: " + e.getMessage());
        } catch (NullPointerException e) {
            mPresenter.logException("Failed to load meta-data, NullPointer: " + e.getMessage());
        }

        TextView appName = root.findViewById(R.id.frag_settings_app_name_tv);
        TextView version = root.findViewById(R.id.frag_settings_version_tv);

        appName.setText(getResources().getString(R.string.app_name));
        version.setText(versionName + getContext().getString(R.string.frag_settings_build) + buildVersion);
    }

    @Override
    public void addPrivacyPolicyInfo() {
        LinearLayout privacyPolicyTv = root.findViewById(R.id.frag_settings_privacy_policy_ll);
        privacyPolicyTv.setOnClickListener(view ->{
            Intent i = new Intent(Intent.ACTION_VIEW);
            i.setData(Uri.parse(view.getContext().getString(R.string.url_privacy_policy)));
            startActivity(i);
        });
    }

    @Override
    public void rateUs() {
        LinearLayout rateUsLL = root.findViewById(R.id.frag_settings_rate_us_ll);
        rateUsLL.setOnClickListener(v -> {
            Uri uri = Uri.parse("market://details?id=" + ApplicationConstants.PACKAGE_NAME);
            Intent intent = new Intent(Intent.ACTION_VIEW,uri);

            // Ignore Playstore backstack, on back press will take us back to our app
            intent.addFlags(Intent.FLAG_ACTIVITY_NO_HISTORY |
                    Intent.FLAG_ACTIVITY_NEW_DOCUMENT |
                    Intent.FLAG_ACTIVITY_MULTIPLE_TASK);

            try{
                startActivity(intent);
            }catch (ActivityNotFoundException e){
                startActivity(new Intent(Intent.ACTION_VIEW,
                        Uri.parse("http://play.google.com/store/apps/details?id=" + ApplicationConstants.PACKAGE_NAME)));
            }
        });

    }

    @Override
    public void setDarkMode() {
        darkModeSwitch = root.findViewById(R.id.frag_settings_dark_mode_switch);
        darkModeSwitch.setChecked(mPresenter.isDarkModeActivated());

        darkModeSwitch.setOnCheckedChangeListener((buttonView, isChecked) -> {
            mPresenter.setDarkMode(isChecked);
            getActivity().recreate();
        });
    }

    public static SettingsFragment newInstance() {
        return new SettingsFragment();
    }

}

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/settings/SettingsFragment.kt
/*
 * The contents of this file are subject to the OpenMRS Public License
 * Version 1.0 (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 * http://license.openmrs.org
 *
 * Software distributed under the License is distributed on an "AS IS"
 * basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See the
 * License for the specific language governing rights and limitations
 * under the License.
 *
 * Copyright (C) OpenMRS, LLC.  All Rights Reserved.
 */
package org.openmrs.mobile.activities.settings

import android.content.ActivityNotFoundException
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.graphics.Typeface
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.AdapterView.OnItemSelectedListener
import android.widget.ArrayAdapter
import android.widget.CompoundButton
import android.widget.TextView
import androidx.fragment.app.viewModels
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import com.google.android.material.snackbar.Snackbar
import com.google.android.material.snackbar.Snackbar.SnackbarLayout
import com.openmrs.android_sdk.utilities.ApplicationConstants
import com.openmrs.android_sdk.utilities.ApplicationConstants.OpenMRSlanguage.LANGUAGE_LIST
import com.openmrs.android_sdk.utilities.ApplicationConstants.ServiceActions.START_CONCEPT_DOWNLOAD_ACTION
import dagger.hilt.android.AndroidEntryPoint
import org.openmrs.mobile.R
import org.openmrs.mobile.activities.BaseFragment
import org.openmrs.mobile.activities.community.contact.ContactUsActivity
import org.openmrs.mobile.activities.logs.LogsActivity
import org.openmrs.mobile.databinding.FragmentSettingsBinding
import org.openmrs.mobile.services.ConceptDownloadService
import org.openmrs.mobile.utilities.ThemeUtils

@AndroidEntryPoint
class SettingsFragment : BaseFragment() {
    private var broadcastReceiver: BroadcastReceiver? = null
    private var _binding: FragmentSettingsBinding? = null
    private val binding get() = _binding!!

    private val viewModel: SettingsViewModel by viewModels()

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        _binding = FragmentSettingsBinding.inflate(inflater, container, false)

        addLogsInfo()
        addBuildVersionInfo()
        addPrivacyPolicyInfo()
        rateUs()
        setupContactUsButton()
        setupDarkMode()
        setupLanguageSpinner()

        broadcastReceiver = object : BroadcastReceiver() {
            override fun onReceive(context: Context, intent: Intent) {
                updateConceptsInDbText()
            }
        }
        setupConceptsView()
        return binding.root
    }

    override fun onPause() {
        super.onPause()
        LocalBroadcastManager.getInstance(requireActivity()).unregisterReceiver(broadcastReceiver!!)
    }

    override fun onResume() {
        super.onResume()
        updateConceptsInDbText()
        LocalBroadcastManager.getInstance(requireActivity())
                .registerReceiver(broadcastReceiver!!, IntentFilter(ApplicationConstants.BroadcastActions.CONCEPT_DOWNLOAD_BROADCAST_INTENT_ID))
    }

    private fun updateConceptsInDbText() {
        val conceptsCount = viewModel.getConceptsCount()
        if (conceptsCount == "0") {
            binding.conceptsDownloadButton.isEnabled = true
            val snackbar = Snackbar.make(binding.root, "", Snackbar.LENGTH_INDEFINITE)
            val customSnackBarView = layoutInflater.inflate(R.layout.snackbar, null)
            val snackBarLayout = snackbar.view as SnackbarLayout
            snackBarLayout.setPadding(0, 0, 0, 0)
            val dismissButton = customSnackBarView.findViewById<TextView>(R.id.snackbar_action_button)
            val typeface = Typeface.createFromAsset(requireActivity().assets, "fonts/Roboto/Roboto-Medium.ttf")
            dismissButton.typeface = typeface
            dismissButton.setOnClickListener { snackbar.dismiss() }
            snackBarLayout.addView(customSnackBarView, 0)
            snackbar.show()
        } else {
            binding.conceptsDownloadButton.isEnabled = false
        }
        binding.conceptsCountTextView.text = conceptsCount
    }

    private fun addLogsInfo() = with(binding) {
        logsDesc1TextView.text = viewModel.logsFileName
        logsDesc2TextView.text = "${getString(R.string.settings_frag_size)} ${viewModel.logSize} kB"
        logsLayout.setOnClickListener {
            startActivity(Intent(context, LogsActivity::class.java))
        }
    }

    private fun setupConceptsView() = with(binding) {
        languageApplyButton.setOnClickListener { requireActivity().recreate() }
        conceptsDownloadButton.setOnClickListener {
            conceptsDownloadButton.isEnabled = false
            Intent(activity, ConceptDownloadService::class.java)
                    .apply { action = START_CONCEPT_DOWNLOAD_ACTION }
                    .let { activity?.startService(it) }
        }
    }

    private fun addBuildVersionInfo() {
        with(binding) {
            appNameTextView.text = getString(R.string.app_name)
            versionTextView.text = viewModel.getBuildVersionInfo(requireContext())
        }
    }

    private fun addPrivacyPolicyInfo() {
        binding.privacyPolicyLayout.setOnClickListener {
            Intent(Intent.ACTION_VIEW)
                    .setData(Uri.parse(getString(R.string.url_privacy_policy)))
                    .let { startActivity(it) }
        }
    }

    private fun rateUs() {
        binding.rateUsLayout.setOnClickListener {
            val intent = Intent(Intent.ACTION_VIEW, viewModel.appMarketUri)
            // Ignore Play Store back stack, on back press will take us back to our app
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                intent.addFlags(Intent.FLAG_ACTIVITY_NO_HISTORY or
                        Intent.FLAG_ACTIVITY_NEW_DOCUMENT or
                        Intent.FLAG_ACTIVITY_MULTIPLE_TASK)
            }
            try {
                startActivity(intent)
            } catch (e: ActivityNotFoundException) {
                startActivity(Intent(Intent.ACTION_VIEW, viewModel.appLinkUri))
            }
        }
    }

    private fun setupContactUsButton() {
        binding.contactUsLayout.setOnClickListener { startActivity(Intent(context, ContactUsActivity::class.java)) }
    }

    private fun setupDarkMode() {
        with(binding) {
            darkModeSwitch.isChecked = ThemeUtils.isDarkModeActivated()
            darkModeSwitch.setOnCheckedChangeListener { _: CompoundButton?, isChecked: Boolean ->
                ThemeUtils.setDarkMode(isChecked)
                requireActivity().recreate()
            }
        }
    }

    private fun setupLanguageSpinner() {
        with(binding.languageSpinner) {
            adapter = ArrayAdapter(requireContext(), android.R.layout.simple_list_item_1, LANGUAGE_LIST)
            setSelection(viewModel.languageListPosition)
            onItemSelectedListener = object : OnItemSelectedListener {
                override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
                    viewModel.languageListPosition = position
                }

                override fun onNothingSelected(parent: AdapterView<*>?) {}
            }
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    companion object {
        fun newInstance(): SettingsFragment {
            return SettingsFragment()
        }
    }
}
